tests/ImageSharp.Benchmarks/PixelBlenders/PorterDuffBulkVsSingleVector.cs

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Six Labors.
+// Licensed under the Six Labors Split License.
+
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+using BenchmarkDotNet.Attributes;
+using SixLabors.ImageSharp.PixelFormats.PixelBlenders;
+
+namespace SixLabors.ImageSharp.Benchmarks.PixelBlenders;
+
+public class PorterDuffBulkVsSingleVector
+{
+    private Vector4[] backdrop;
+    private Vector4[] source;
+
+    [GlobalSetup]
+    public void Setup()
+    {
+        this.backdrop = new Vector4[8 * 20];
+        this.source = new Vector4[8 * 20];
+
+        FillRandom(this.backdrop);
+        FillRandom(this.source);
+    }
+
+    private static void FillRandom(Vector4[] arr)
+    {
+        Random rng = new();
+        for (int i = 0; i < arr.Length; i++)
+        {
+            arr[i].X = rng.NextSingle();
+            arr[i].Y = rng.NextSingle();
+            arr[i].Z = rng.NextSingle();
+            arr[i].W = rng.NextSingle();
+        }
+    }
+
+    [Benchmark(Description = "Scalar")]
+    public Vector4 OverlayValueFunction_Scalar()
+    {
+        Vector4 result = default;
+        for (int i = 0; i < this.backdrop.Length; i++)
+        {
+            result = PorterDuffFunctions.NormalSrcOver(this.backdrop[i], this.source[i], .5F);
+        }
+
+        return result;
+    }
+
+    [Benchmark(Description = "Avx")]
+    public Vector256<float> OverlayValueFunction_Avx()
+    {
+        ref Vector256<float> backdrop = ref Unsafe.As<Vector4, Vector256<float>>(ref MemoryMarshal.GetReference<Vector4>(this.backdrop));
+        ref Vector256<float> source = ref Unsafe.As<Vector4, Vector256<float>>(ref MemoryMarshal.GetReference<Vector4>(this.backdrop));
+
+        Vector256<float> result = default;
+        Vector256<float> opacity = Vector256.Create(.5F);
+        int count = this.backdrop.Length / 2;
+        for (int i = 0; i < count; i++)
+        {
+            result = PorterDuffFunctions.NormalSrcOver(Unsafe.Add(ref backdrop, i), Unsafe.Add(ref source, i), opacity);
+        }
+
+        return result;
```

Comment:
Is there a bug in my benchmark? 

Numbers seem to good to be true.

```
BenchmarkDotNet=v0.13.0, OS=Windows 10.0.22621
11th Gen Intel Core i7-11370H 3.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK=7.0.102
  [Host]     : .NET 6.0.14 (6.0.1423.7309), X64 RyuJIT
  DefaultJob : .NET 6.0.14 (6.0.1423.7309), X64 RyuJIT


| Method |       Mean |   Error |  StdDev |
|------- |-----------:|--------:|--------:|
| Scalar | 2,176.5 ns | 4.48 ns | 3.97 ns |
|    Avx |   231.3 ns | 0.87 ns | 0.82 ns |

// * Hints *
Outliers
  PorterDuffBulkVsSingleVector.Scalar: Default -> 1 outlier  was  removed (2.20 us)

// * Legends *
  Mean   : Arithmetic mean of all measurements
  Error  : Half of 99.9% confidence interval
  StdDev : Standard deviation of all measurements
  1 ns   : 1 Nanosecond (0.000000001 sec)

// ***** BenchmarkRunner: End *****
// ** Remained 0 benchmark(s) to run **
Run time: 00:00:26 (26.39 sec), executed benchmarks: 2

Global total time: 00:00:47 (47.57 sec), executed benchmarks: 2
// * Artifacts cleanup *
```

Reply:
I notice that both `backdrop` and `source` references `this.backdrop`?

Reply:
Good catch! Doesn't affect the results though.

Reply:
Just peaking at .NET 7 and changing the vector algorithm to be `V128` to simplify the diff a bit, then we have:

### Vector4
```asm
; Assembly listing for method C:NormalSrcOver(System.Numerics.Vector4,System.Numerics.Vector4,float):System.Numerics.Vector4
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; optimized code
; rsp based frame
; partially interruptible
; No PGO data
; 0 inlinees with PGO data; 3 single block inlinees; 0 inlinees without PGO data

G_M000_IG01:                ;; offset=0000H
       4883EC18             sub      rsp, 24
       C5F877               vzeroupper 
       C5F8293424           vmovaps  xmmword ptr [rsp], xmm6
 
G_M000_IG02:                ;; offset=000CH
       498D400C             lea      rax, bword ptr [r8+0CH]
       C5E25900             vmulss   xmm0, xmm3, dword ptr [rax]
       C5FA1100             vmovss   dword ptr [rax], xmm0
       C4C1781000           vmovups  xmm0, xmmword ptr [r8]
       C5F828C8             vmovaps  xmm1, xmm0
       C5F81012             vmovups  xmm2, xmmword ptr [rdx]
       C5E8C6DAFF           vshufps  xmm3, xmm2, xmm2, -1
       C5F0C6E1FF           vshufps  xmm4, xmm1, xmm1, -1
       C5E259EC             vmulss   xmm5, xmm3, xmm4
       C5E25CDD             vsubss   xmm3, xmm3, xmm5
       C5DA5CF5             vsubss   xmm6, xmm4, xmm5
       C5E258E4             vaddss   xmm4, xmm3, xmm4
       C4E27918DB           vbroadcastss xmm3, xmm3
       C5E859D3             vmulps   xmm2, xmm2, xmm3
       C4E27918DE           vbroadcastss xmm3, xmm6
       C5F059CB             vmulps   xmm1, xmm1, xmm3
       C5E858C9             vaddps   xmm1, xmm2, xmm1
       C4E27918D5           vbroadcastss xmm2, xmm5
       C5F859C2             vmulps   xmm0, xmm0, xmm2
       C5F058C0             vaddps   xmm0, xmm1, xmm0
       C5F828CC             vmovaps  xmm1, xmm4
       C5F8101532000000     vmovups  xmm2, xmmword ptr [reloc @RWD00]
       C5E85FC9             vmaxps   xmm1, xmm2, xmm1
       C4E27918C9           vbroadcastss xmm1, xmm1
       C5F85EC1             vdivps   xmm0, xmm0, xmm1
       C4E37921C430         vinsertps xmm0, xmm0, xmm4, 48
       C5F81101             vmovups  xmmword ptr [rcx], xmm0
       488BC1               mov      rax, rcx
 
G_M000_IG03:                ;; offset=0088H
       C5F8283424           vmovaps  xmm6, xmmword ptr [rsp]
       4883C418             add      rsp, 24
       C3                   ret   
```

### Vector128

```asm
G_M000_IG01:                ;; offset=0000H
       4883EC18             sub      rsp, 24
       C5F877               vzeroupper 
       C5F8293424           vmovaps  qword ptr [rsp], xmm6
 
G_M000_IG02:                ;; offset=000CH
       C4C1791000           vmovupd  xmm0, xmmword ptr [r8]
       C4C1785909           vmulps   xmm1, xmm0, xmmword ptr [r9]
       C4E3790CC188         vblendps xmm0, xmm0, xmm1, -120
       C4C1791100           vmovupd  xmmword ptr [r8], xmm0
       C4C1791000           vmovupd  xmm0, xmmword ptr [r8]
       C5F828C8             vmovaps  xmm1, xmm0
       C5F91012             vmovupd  xmm2, xmmword ptr [rdx]
       C4E37904D8FF         vpermilps xmm3, xmm0, -1
       C4E37904E2FF         vpermilps xmm4, xmm2, -1
       C5E059EC             vmulps   xmm5, xmm3, xmm4
       C5D85CE5             vsubps   xmm4, xmm4, xmm5
       C5E05CF5             vsubps   xmm6, xmm3, xmm5
       C5D858DB             vaddps   xmm3, xmm4, xmm3
       C5E859D4             vmulps   xmm2, xmm2, xmm4
       C4E279B8D6           vfmadd231ps xmm2, xmm0, xmm6
       C4E271B8D5           vfmadd231ps xmm2, xmm1, xmm5
       C5E05F0530000000     vmaxps   xmm0, xmm3, xmmword ptr [reloc @RWD00]
       C5E85ED0             vdivps   xmm2, xmm2, xmm0
       C4E3690CC388         vblendps xmm0, xmm2, xmm3, -120
       C5F91101             vmovupd  xmmword ptr [rcx], xmm0
       488BC1               mov      rax, rcx
 
G_M000_IG03:                ;; offset=0071H
       C5F8283424           vmovaps  xmm6, qword ptr [rsp]
       4883C418             add      rsp, 24
       C3                   ret
```

The `V256` code is effectively identical to the `V128` code, just using `ymm` instead of `xmm` and we can assume is roughly 2x the speed.

For the differences between `Vector4` and `V128`...

Reply:
The initial `source.W *= opacity` isn't great, it's generating:

```asm
lea      rax, bword ptr [r8+0CH]
vmulss   xmm0, xmm3, dword ptr [rax]
vmovss   dword ptr [rax], xmm0
vmovupd  xmm0, xmmword ptr [r8]
```

If `NormalSrcOver` was inlined and `source` was in register, it'd also be generating a spill before this:
```asm
vmovapd  xmmword ptr [r8], xmm0
```

That basically means it's doing:
* write source to memory
* get the address of `w`
* load `w` from memory and multiply by `opacity`
* store result into `w`
* load entire vector from memory

Compare this with the `V128` approach:
```asm
vmovupd  xmm0, xmmword ptr [r8]
vmulps   xmm1, xmm0, xmmword ptr [r9]
vblendps xmm0, xmm0, xmm1, 8
```

Which is:
* load source from memory
* load opacity from memory and multiply by source
* blend so that you get original source plus new W

If you instead did something roughly equivalent, making it `Vector4 opacity` and then did `source.W = (source * opacity).W;`, you get this instead:
```csharp
vmovupd  xmm0, xmmword ptr [r8]
vmulps   xmm1, xmm0, xmmword ptr [r9]
vshufps  xmm1, xmm1, xmm1, -1
vinsertps xmm0, xmm0, xmm1, 48
```

Which is nearly ideal and doing the same thing, there's just a stray `vshufps` that I need to remove on the JIT side >.>

Reply:
The next 6 instructions are "the same" between the two algorithms, but there is a slight codegen pessimization from the JIT:

We have for `Vector4`:
```asm
vmovupd  xmmword ptr [r8], xmm0
vmovupd  xmm0, xmmword ptr [r8]
vmovaps  xmm1, xmm0
vmovupd  xmm2, xmmword ptr [rdx]
vshufps  xmm3, xmm2, xmm2, -1
vshufps  xmm4, xmm1, xmm1, -1
```

and for `V128`:
```asm
vmovupd  xmmword ptr [r8], xmm0
vmovupd  xmm0, xmmword ptr [r8]
vmovaps  xmm1, xmm0
vmovupd  xmm2, xmmword ptr [rdx]
vpermilps xmm3, xmm0, -1
vpermilps xmm4, xmm2, -1
```

They really just differ on whether `shuffle` or `permute` is picked. There is an initial spill and load of `source` which shouldn't be necessary, but the JIT is introducing due to some complexities around struct returns on Windows (it's caused by `Normal(backdrop, source)` being a "nop" effectively) and also due to required evaluation order of passed parameters.

There's not anything for you to do here, just describing the difference.

Reply:
The last bit primarily comes down to the `Vector4` version again pulling out and operating on scalars, rather than treating it as a "single vector". There is also a difference in whether `Fma(a, b, c)` or `Add(Multiply(a, b), c)` is used, but that's "less important" and could still be handled with a tiny helper method.

Changing the `Vector4` implementation to more closely match `V128`, gives us:
```csharp
Vector4 sW = new Vector4(source.W);
Vector4 dW = new Vector4(destination.W);

// calculate weights
Vector4 blendW = sW * dW;
Vector4 dstW = dW - blendW;
Vector4 srcW = sW - blendW;

// calculate final alpha
Vector4 alpha = dstW + sW;

// calculate final color
Vector4 color = (destination * dstW);
color = MultiplyAdd(color, source, srcW);
color = MultiplyAdd(color, blend, blendW);

// unpremultiply
color /= Vector4.Max(alpha, new Vector4(Epsilon));
color.W = alpha.W;
return color;
```

This generates:
```asm
C5F0C6D9FF           vshufps  xmm3, xmm1, xmm1, -1
C4E27918DB           vbroadcastss xmm3, xmm3
C5E8C6E2FF           vshufps  xmm4, xmm2, xmm2, -1
C4E27918E4           vbroadcastss xmm4, xmm4
C5E059EC             vmulps   xmm5, xmm3, xmm4
C5D85CE5             vsubps   xmm4, xmm4, xmm5
C5E05CF5             vsubps   xmm6, xmm3, xmm5
C5D858DB             vaddps   xmm3, xmm4, xmm3
C5E859D4             vmulps   xmm2, xmm2, xmm4
C5F059CE             vmulps   xmm1, xmm1, xmm6
C5F058D2             vaddps   xmm2, xmm1, xmm2
C5F859C5             vmulps   xmm0, xmm0, xmm5
C5F858D2             vaddps   xmm2, xmm0, xmm2
C4E27918053C000000   vbroadcastss xmm0, dword ptr [reloc @RWD00]
C5E05FC0             vmaxps   xmm0, xmm3, xmm0
C5E85ED0             vdivps   xmm2, xmm2, xmm0
C5E0C6C3FF           vshufps  xmm0, xmm3, xmm3, -1
C4E36921D030         vinsertps xmm2, xmm2, xmm0, 48
C5F91111             vmovupd  xmmword ptr [rcx], xmm2
488BC1               mov      rax, rcx
```

This is quite a bit better, but still not "ideal". There are notably some unnecessary `broadcastss` after the first two `shufps` and an unnecessary `shufps` before the last `insertps` (same general thing as mentioned for the first scenario covering `source.W = (source * opacity).W`).

For those couple of remaining pessimizations there's nothing that ImageSharp could "improve" itself there for `Vector4`, its codegen fixes that the JIT will need to make for .NET 8

Reply:
It could however "improve" it if it wanted.

Notably it could selectively use `vector4.AsVector128()` and `vector128.AsVector4()` to convert between `Vector4` and `Vector128<float>`. It could then specially handle these couple edge cases with small helper methods.

For example, you could define a helper:
```csharp
private static Vector4 PermuteW(Vector4 value)
{
    return Sse.Shuffle(value.AsVector128(), value.AsVector128(), 0xFF).AsVector4();
}
```

and then change the following:
```diff
-Vector4 sW = new Vector4(source.W); 
-Vector4 dW = new Vector4(destination.W);
+Vector4 sW = PermuteW(source);
+Vector4 dW = PermuteW(destination);
```

That then removes the two unnecessary `broadcastss`.

You could likewise create a helper for:
```csharp
private static Vector4 WithW(Vector4 value, Vector4 w)
{
    // TODO: Provide SSE fallback which uses "shuffle" - just pick XYZ from value and W from w
    return Sse41.Insert(value.AsVector128(), w.AsVector128(), 0b11_11_0000).AsVector4();
}
```

and do the following diff:
```diff
-source.W = (source * opacity).W;
+source = WithW(source, source * opacity);
```

also:
```diff
-color.W = alpha.W;
+color = WithW(color, alpha);
```

Reply:
I'll make sure JIT issues are logged for the couple pessimizations on the JIT side so that manual workarounds aren't needed long term.

Reply:
Insanely detailed writeup, learnt a lot, thank you!

Reply:
This is incredible  @tannergooding Thanks so much!

I'm gonna have a tinker and see where I can get. `NormalSrcOver` is currently 9.2X faster using AVX which is absolutely bonkers. CC @antonfirsov 

```
BenchmarkDotNet=v0.13.0, OS=Windows 10.0.22621
11th Gen Intel Core i7-11370H 3.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK=7.0.102
  [Host]     : .NET 6.0.14 (6.0.1423.7309), X64 RyuJIT
  DefaultJob : .NET 6.0.14 (6.0.1423.7309), X64 RyuJIT


| Method |       Mean |   Error |  StdDev | Ratio |
|------- |-----------:|--------:|--------:|------:|
| Scalar | 2,177.0 ns | 4.38 ns | 3.66 ns |  1.00 |
|    Avx |   233.5 ns | 1.25 ns | 1.17 ns |  0.11 |
```

Reply:
MAJOR improvement for the Vector4 path. @tannergooding

I implemented the `WithW` and the things immediate dropped to nearly 4X faster. `PermuteW` brought it down even further.

The only thing I couldn't figure out was how to fallback to `Sse` for `WithW`. I'm not great with shuffling.

```
BenchmarkDotNet=v0.13.0, OS=Windows 10.0.22621
11th Gen Intel Core i7-11370H 3.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK=7.0.102
  [Host]     : .NET 6.0.14 (6.0.1423.7309), X64 RyuJIT
  DefaultJob : .NET 6.0.14 (6.0.1423.7309), X64 RyuJIT


| Method |     Mean |   Error |  StdDev | Ratio |
|------- |---------:|--------:|--------:|------:|
| Scalar | 448.6 ns | 0.99 ns | 0.88 ns |  1.00 |
|    Avx | 235.1 ns | 1.23 ns | 1.03 ns |  0.52 |
```

Reply:
> MAJOR improvement for the Vector4 path.

Awesome! Much closer to the 1.5-2x difference we'd otherwise expect

> The only thing I couldn't figure out was how to fallback to Sse for WithW.

Think you want:
```
Vector128<float> tmp = Sse.Shuffle(w.AsVector128(), value.AsVector128(), 0b00_10_00_11);
return Sse.Shuffle(value.AsVector128(), tmp, 0b00_10_01_00);
```

That will create `tmp` as `<w[3], w[0], value[2], value[0]>`
Then return `<value[0], value[1], tmp[2], tmp[0]>` (which is `<value[0], value[1], value[2], w[3]>`)

Reply:
I would not have figured that one out. Thanks!